clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -274,10 +273,10 @@ public ReadyCheckResult ready(Cluster cluster, long nowMs) {
             Deque<RecordBatch> deque = entry.getValue();
 
             Node leader = cluster.leaderFor(part);
-            if (leader == null) {
-                unknownLeadersExist = true;
-            } else if (!readyNodes.contains(leader)) {
-                synchronized (deque) {
+            synchronized (deque) {
+                if (leader == null && !deque.isEmpty()) {
```

Comment:
@ewencp Doesn't the normal expiry of batches in `RecordAccumulator.abortExpiredBatches` handle removal of entries, preventing the indefinite growth of `this.batches`?